clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -376,39 +376,58 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
-                have_cond_branch_target = true;
-                cond_branch_target = cached->second;
-            } else {
-                instr_t instr;
-                instr_init(GLOBAL_DCONTEXT, &instr);
-                const app_pc decode_pc =
-                    const_cast<app_pc>(shard->prev_instr_.instr.encoding);
-                const app_pc next_pc =
-                    decode_from_copy(GLOBAL_DCONTEXT, decode_pc,
-                                     reinterpret_cast<app_pc>(trace_pc), &instr);
-                if (next_pc == nullptr || !opnd_is_pc(instr_get_target(&instr))) {
-                    // Neither condition should happen but they could on an invalid
-                    // encoding from raw2trace or the reader so we report an
-                    // invariant rather than asserting.
-                    report_if_false(shard, false, "Branch target is not decodeable");
-                } else {
+            instr_t instr;
+            instr_init(GLOBAL_DCONTEXT, &instr);
+            const app_pc decode_pc =
+                const_cast<app_pc>(shard->prev_instr_.instr.encoding);
+            const app_pc next_pc = decode_from_copy(
+                GLOBAL_DCONTEXT, decode_pc, reinterpret_cast<app_pc>(trace_pc), &instr);
+            if (type_is_instr_direct_branch(shard->prev_instr_.instr.type)) {
+                if (shard->prev_instr_.instr.encoding_is_new)
+                    shard->branch_target_cache.erase(trace_pc);
+                auto cached = shard->branch_target_cache.find(trace_pc);
+                if (cached != shard->branch_target_cache.end()) {
                     have_cond_branch_target = true;
-                    cond_branch_target =
-                        reinterpret_cast<addr_t>(opnd_get_pc(instr_get_target(&instr)));
-                    shard->branch_target_cache[trace_pc] = cond_branch_target;
+                    cond_branch_target = cached->second;
+                } else {
+                    if (next_pc == nullptr || !opnd_is_pc(instr_get_target(&instr))) {
+                        // Neither condition should happen but they could on an invalid
+                        // encoding from raw2trace or the reader so we report an
+                        // invariant rather than asserting.
+                        report_if_false(shard, false, "Branch target is not decodeable");
+                    } else {
+                        have_cond_branch_target = true;
+                        cond_branch_target = reinterpret_cast<addr_t>(
+                            opnd_get_pc(instr_get_target(&instr)));
+                        shard->branch_target_cache[trace_pc] = cond_branch_target;
+                    }
                 }
-                instr_free(GLOBAL_DCONTEXT, &instr);
+            } else {
+                instr_t cur_instr;
+                instr_init(GLOBAL_DCONTEXT, &cur_instr);
+                app_pc curr_pc = decode_from_copy(
```

Comment:
We still have multiple code snippets that are decoding instrs. Like I mentioned in previous review, in each iteration we should be decoding only the cur instr (to reduce decoding overhead) and store it in the `shard` for use in the next iteration.